### PR TITLE
Pla 7111/mean features builder

### PR DESCRIPTION
**Pull request by bfolie**
# Citrine Python PR

## Description 
Creates a Builder method to help a user calculate mean properties where the properties are calculated by a featurizer. For example, the user might have SMILES strings for a formulation's ingredients, and want to calculate the mean of the molecular structure features computed based on the SMILES strings.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [X] I have added tests for 100% coverage
- [X] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [X] I have bumped the version in setup.py

### CODE REVIEW

File: src/citrine/builders/predictors.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    if isinstance(featurizer, MolecularStructureFeaturizer):
+        input_descriptor = featurizer.descriptor
+    else:
+        raise TypeError(f"Featurizer of type {type(featurizer)} is not supported.")
+
+    if labels is None:
+        labels = []
+    if all_ingredients:
+        labels = [None] + labels
+    if not isinstance(labels, list) or len(labels) == 0:
```
Comment (latture) [inline src/citrine/builders/predictors.py:68]: Based on the error message, I would have expected the following code:

```suggestion
    if not all_ingredients or len(labels) == 0:
```

Comment (bfolie) [reply src/citrine/builders/predictors.py:68]: It's not an exclusive or. I'll change the message. This logic is doing two things -- it's making sure that `all_ingredients == true` or labels are specified, and it's also make sure that `labels` was specified as a list and not something else.

File: src/citrine/builders/predictors.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    """
+    if isinstance(featurizer, MolecularStructureFeaturizer):
+        input_descriptor = featurizer.descriptor
+    else:
+        raise TypeError(f"Featurizer of type {type(featurizer)} is not supported.")
+
+    if labels is None:
+        labels = []
+    if not isinstance(labels, list):
+        raise TypeError(f"labels must be specified as a list of strings")
```
Comment (latture) [inline src/citrine/builders/predictors.py:67]:
```suggestion
        raise TypeError("labels must be specified as a list of strings")
```

File: src/citrine/builders/predictors.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        accepts a molecular structure featurizer, but other models can be used in principle.
+        TODO: add chemical formula featurizer once it is exposed
+    formulation_descriptor: FormulationDescriptor
+        Descriptor that represents the formulation being featurized.
+    p: int
+        Power of the generalized mean. Only integer powers are supported.
+    impute_properties: bool
+        Whether to impute missing ingredient properties by averaging over the entire dataset.
+        If ``False``, an error is thrown if any ingredient being featurized is missing a property.
+    all_ingredients: bool
```
Comment (latture) [inline src/citrine/builders/predictors.py:44]: Maybe rename this to `featurize_all_ingredients` or something similar? I'm not sure what it's doing to `all_ingredients`. Also maybe describe what happens if it's false: `If False, MPP's are only constructed for the specified labels. Must be True if no Labels are specified.`

File: src/citrine/builders/predictors.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        labels = [None] + labels
+    if len(labels) == 0:
+        msg = "No mean property predictors requested. " \
+              "Set all_ingredients to True and/or specify labels."
+        raise ValueError(msg)
+
+    properties = project.descriptors.from_predictor_responses(featurizer, [input_descriptor])
+    real_properties = [desc for desc in properties if isinstance(desc, RealDescriptor)]
+    if len(real_properties) == 0:
+        raise RuntimeError("")
```
Comment (latture) [inline src/citrine/builders/predictors.py:78]: Missing error message?

File: src/citrine/builders/predictors.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        p: int,
+        impute_properties: bool,
+        label: Optional[str]
+) -> MeanPropertyPredictor:
+    """Build a MeanPropertyPredictor for given specifications."""
+    name = f"mean of {ingredient_descriptor.key} features"
+    if p != 1:
+        name = f"{p}-" + name
+    if label is not None:
+        name = name + f" for label {label}"
```
Comment (latture) [inline src/citrine/builders/predictors.py:117]:
```suggestion
        name += f" for label {label}"
```

File: src/citrine/builders/predictors.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    name = name + f" in {formulation_descriptor.key}"
```
Comment (latture) [inline src/citrine/builders/predictors.py:118]:
```suggestion
    name += f" in {formulation_descriptor.key}"
```

File: src/citrine/builders/predictors.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        formulation_descriptor: FormulationDescriptor,
+        properties: List[RealDescriptor],
+        p: int,
+        impute_properties: bool,
+        label: Optional[str]
+) -> MeanPropertyPredictor:
+    """Build a MeanPropertyPredictor for given specifications."""
+    name = f"mean of {ingredient_descriptor.key} features"
+    if p != 1:
+        name = f"{p}-" + name
```
Comment (latture) [inline src/citrine/builders/predictors.py:115]:
```suggestion
        name = f"{p}-{name}"
```

**Review by latture — CHANGES_REQUESTED** [review]
Overall, the logic and tests look great. Only a few minor comments/suggestions, and looks like there's a missing error message.

File: src/citrine/builders/predictors.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+from citrine.resources.project import Project
+from citrine.informatics.predictors import (
+    MolecularStructureFeaturizer,
+    MeanPropertyPredictor
+)
+from citrine.informatics.descriptors import Descriptor, FormulationDescriptor, RealDescriptor
+
+
+def mean_feature_properties(
```
[the remark this answers is not in the document]
Comment (bfolie) [reply src/citrine/builders/predictors.py:11]: good point; I'll change it

File: src/citrine/builders/predictors.py (hunk 2 above)
Comment (bfolie) [reply src/citrine/builders/predictors.py:67]: > Also, why not allow sets?

🤷

File: src/citrine/builders/predictors.py (hunk 4 above)
Comment (bfolie) [reply src/citrine/builders/predictors.py:78]: 🤦

**Review by latture — DISMISSED** [review]
Looks good to me.

**Review by maxhutch — APPROVED** [review]

**Pull request merged**